Template-level subscriptions

stubailo opened the pull request:
Now, you can call `this.subscribe` inside the `onCreated` callback of a Template. This subscription is automatically stopped when the template is destroyed. You also get a bonus built-in helper, `Template.ready`, that returns `true` when all of the subscriptions are ready. You can also manage your ready state independently by using the handles returned from `this.subscribe`.

stubailo commented:
Added tests and documentation for the new feature.

stubailo commented:
@avital @dgreensp review please?

avital commented on packages/spacebars-tests/template_tests.js:
```diff
@@ -3067,3 +3067,38 @@ Tinytest.add("spacebars-tests - template_tests - inclusion with data remove (#31
+
+    tmpl.onCreated(function () {
+      subHandle = this.subscribe("items", subscribeCallback);
+      subHandle.stop = stopCallback;
```
That's a nice idea -- basically mock out the `stop` method instead of testing for the results of it being called. I'll use this in other tests!

avital commented on packages/spacebars-tests/template_tests.js:
```diff
@@ -3067,3 +3067,38 @@ Tinytest.add("spacebars-tests - template_tests - inclusion with data remove (#31
+    });
+
+    tmpl.onCreated(function () {
+      subHandle = this.subscribe("items", subscribeCallback);
```
Maybe have more than one subscription? So that you test the behavior of "they all get stopped", and the behavior of "only if they are all ready is `Template.ready` true"

stubailo replied:
Added a test for multiple subs!

avital commented on packages/spacebars-tests/template_tests.js:
```diff
@@ -3067,3 +3067,38 @@ Tinytest.add("spacebars-tests - template_tests - inclusion with data remove (#31
+    // Manually check the subscribe ready callback to make sure the template is
+    // doing the right thing
+    var subscribeCallback = expect(function () {
+      Tracker.afterFlush(checkHTMLThenRemove);
```
I think you can simplify this by instead calling `Tracker.flush()`. Which would get rid of one layer of `expect` and function definitions.

stubailo replied:
For some reason, it doesn't work if I do that.

dgreensp replied:
Maybe `subscribe` is using `afterFlush` itself.

avital commented on packages/blaze/view.js:
```diff
@@ -184,6 +184,34 @@ Blaze.View.prototype.autorun = function (f, _inViewScope) {
+  var self = this;
+
+  if (! self.isCreated) {
+    throw new Error("View#subscribe must be called from the created callback at the earliest");
```
Are there additional limitations? Like -- a user probably shouldn't call this from the destroyed callback?

And if it's OK to call this from the rendered callback I think it would be good to test that explicitly (in the same test if needed, just add an additional subscription from a rendered callback)

avital commented on packages/blaze/view.js:
```diff
@@ -184,6 +184,34 @@ Blaze.View.prototype.autorun = function (f, _inViewScope) {
+  if (! self.isCreated) {
+    throw new Error("View#subscribe must be called from the created callback at the earliest");
+  }
+  if (this._isInRender) {
```
If I understand correctly, these two if clauses (and the one above I guess) are really an attempt to say: "if not in a created or rendered callback throw an error". Is there a better way to check for this? (I don't know the answer). It looks  a little ad-hoc.

At the very least I'd recommend extracting these into a utility function called, say, "canCallSubscribe" that's documented as checking if you're in a created or rendered callback with a comment saying this is the best way you found to check for this.

Maybe @dgreensp has a better suggestion.

dgreensp replied:
This is a typical pattern in the Blaze code, but maybe it should be moved into a helper.

stubailo replied:
Yeah I copy-pasted this from `this.autorun` :]

Definitely should be refactored a little

avital commented:
Added some comments, but this looks great! So simple!

arunoda commented:
Very nice :)

dgreensp commented on packages/spacebars-compiler/codegen.js:
```diff
@@ -27,7 +27,9 @@ var builtInTemplateMacros = {
-  'dynamic': 'Template.__dynamic'
+  'dynamic': 'Template.__dynamic',
+
+  'ready': 'view.templateInstance().ready()'
```
If `view` isn't a template, this will throw an error that `view.templateInstance` is undefined, right?  Shouldn't we throw a better error?

stubailo replied:
I think that is impossible because this code can only occur inside of a template.

dgreensp replied:
Oh, you're right, `view` is actually the template's view.

dgreensp commented on packages/blaze/template.js:
```diff
@@ -309,6 +311,43 @@ Blaze.TemplateInstance.prototype.autorun = function (f) {
+ * @return {Boolean} True if all subscriptions on this template instance are
+ * ready.
+ */
+Blaze.TemplateInstance.prototype.ready = function () {
```
This could break code that uses a `ready` property on the template instance.  It doesn't break the Discover Meteor code, which sets `this.ready = new ReactiveVar` on the template instance (because the inherited property will be shadowed), but it would break someone's code who checks `this.ready` but never explicitly initializes it to `false` (since now `this.ready` starts out truthy).

I think I'm ok with that, but we should at least have a unit test that verifies that shadowing `this.ready()` works.  (I don't see any tests of `this.ready()` actually, only `Template.ready`.)  That way, we won't accidentally break apps that rely on being able to set `this.ready` (like apps following Discover Meteor's pattern) at some point in the future if we change how the template instance object works.

If someone shadows `this.ready`, I also note that it will break `Template.ready`.  But we could decide we don't care about that.

dgreensp replied:
And we should mention it wherever we mention breaking changes.

It could turn out that a lot of people set `this.ready` in their apps and are unclear on whether they need to change that code now or not.

stubailo replied:
After talking to matt, will change this to `subscriptionsReady`

stubailo commented:
Currently in the middle of implementing onStopped for subscriptions, put this CR on hold please

avital commented on packages/spacebars-tests/template_tests.js:
```diff
@@ -3067,3 +3067,59 @@ Tinytest.add("spacebars-tests - template_tests - inclusion with data remove (#31
+      }
+
+      if (subscriptionsFinished === 2) {
+        Tracker.afterFlush(checkBothReady);
```
(Lost the original comment since you re-wrote the test)

Before we merge this, I'd like to understand why we can't use `Deps.flush`. (Happy to pair on it if that would be helpful)

stubailo replied:
No progress on this. Do you want to pair on it?

Tarang commented:
Is it not possible to allow a reactive change with this?

i.e
```
Template.hello.onCreated(function() {
    var self = this;
    self.autorun(function() {
      self.subscribe("list", Session.get('category'))
    });
  })
```

stubailo commented:
@Tarang the intention is that the code you provided will work as expected. I should add a test for that particular use case.

stubailo commented:
Rebased on devel and fixed minor merge conflicts. @glasser, what do you think?

In particular, I have made two changes to subscriptions:

1. The `onError` callback is deprecated now, and replaced with `onStopped` with an optional error callback. I still have to write docs for this change, but it's fully backwards compatible.
2. Meteor.subscribe inside an autorun always returns the exact same handle, so you can compare them with `===`.

stubailo commented:
OK all of the tests pass and I have added a lot of new tests. @glasser, @avital what else should I do before merging this? Also, should I rebase some things to reduce the number of commits?

glasser commented on docs/client/full-api/api/pubsub.md:
```diff
@@ -173,10 +173,11 @@ collection name.
+The `onReady` callback is called with no arguments when the server [marks the
+subscription as ready](#publish_ready). The `onStopped` callback is called with
+a [`Meteor.Error`](#meteor_error) if the subscription fails or is terminated by
+the server. If the subscription is stopped for any other reason, `onStopped` is
```
we can be specific --- stopped by calling `stop` on either the client or the server (right?)

are there other ways?  hmm, i actually have no idea what happens when you close a DDP connection (with the not-really-documented `connection.close()` --- does it stop the subscriptions and remove their contents from minimongo?  guessing not, though that is actually kinda weird.

glasser replied:
also i'm not sure if we should still document onError too (as deprecated). I guess we usually don't.  but it does make it harder for people to read slightly old code.  well, as long as History.md is updated....

stubailo replied:
I don't think we have a history of documenting deprecated things. There's a general issue here where it's really hard to access the docs of old meteor versions. Maybe it's worth uploading all of the old docs apps with URLs like "docs-1.0.3.1.meteor.com" or something

glasser replied:
Ah yes I guess I meant --- What you are doing here is totally the way we do things here.  I'm starting to think it's not ideal for users, though.

It wouldn't actually be all that hard to upload all the old docs apps...

glasser commented on packages/ddp/livedata_connection.js:
```diff
@@ -1406,11 +1436,25 @@ _.extend(Connection.prototype, {
     self._subscriptions[msg.id].remove();
+
+    var meteorErrorFromMsg = function (msgArg) {
+      return msgArg && new Meteor.Error(
```
i feel like this is missing a check on `msgArg.error`

stubailo replied:
Done!

glasser commented on packages/blaze/template.js:
```diff
@@ -267,6 +267,8 @@ Blaze.TemplateInstance = function (view) {
    * @type {DOMNode}
```
Is it possible to separate this into one commit which adds the feature to ddp (and doesn't touch blaze at all) followed by one commit that does the blaze/spacebars stuff?

eg, let's say we find a bug in the blaze thing and revert it.  would be a shame to lose the new feature.

if you need help splitting git commits in two i can help :)

stubailo replied:
This is the commit you want to split? https://github.com/meteor/meteor/commit/1f77d8e40ffca2f9b86e9fb5078b6c77487acb32

glasser commented on packages/ddp/livedata_connection.js:
```diff
@@ -520,12 +532,41 @@ _.extend(Connection.prototype, {
+        existing.stoppedCallback = callbacks.onStopped;
+      }
+
+      // Return the exact same handle so that we can use === to check if the
```
interesting.  where are you using this?  this seems like a strong assumption that, if we want to make it dependable, should be documented.

stubailo replied:
In order to stop all of the subscriptions when the template is destroyed, I need to keep a list of all of the handles. To stop that list from growing without bound, I need to be able to deduplicate them. My two options were either using `===` to compare handles or having the handle include the subscription id. I thought not leaking the ID was the less risky option.

glasser replied:
That makes sense, but in that case it should probably be documented, at least internally.  At the very least, the (good) comment that you added here could have a link to the blaze code that cares.  Or it could even be in the real docs (which does describe the de-duplication), though that links would still be good.

On the other hand --- the `id` is an actual real part of the DDP wire protocol, not an internal implementation detail of the ddp package, and arguably we should just expose it on the handle?

stubailo replied:
I'm open to either one. It seems weird to generate lots of identical handles on the one hand, on the other hand the `id` solution could be less magical and easier to document. Maybe we should do both? Basically don't generate lots of unnecessary handles as an optimization but also put the `id` on there, document that change, and use it everywhere.

What do you think?

stubailo replied:
On the other hand it will make code like `_.union(subHandles.get(), [subHandle])` less nice to start using the `subscriptionId`.

glasser replied:
well, i mean, just use a map instead of an array?  then it actually becomes easier?  (or even ReactiveDict instead of ReactiveVar if that makes sense)

glasser replied:
(though you can't iterate over a reactivedict... see #3135)

glasser commented on packages/ddp/livedata_tests.js:
```diff
@@ -601,29 +611,109 @@ if (Meteor.isClient) {
     ];})());
 
-    testAsyncMulti("livedata - publish multiple cursors", [
+  testAsyncMulti("livedata - publisher errors with onStopped callback", (function () {
```
copy and pasting the whole test is not really ideal (though the ugly thing we do where we make tests that run in multiple guises isn't super great either) but better than no test

stubailo replied:
I guess I could do a for loop and define some function like "onStoppedPretendingToBeOnError" but this seemed easier to understand.

glasser commented:
I didn't look at the blaze side of things, but finished a pass on the ddp side

glasser commented on packages/blaze/template.js:
```diff
@@ -313,6 +315,78 @@ Blaze.TemplateInstance.prototype.autorun = function (f) {
+ * @return {Boolean} True if all subscriptions on this template instance are
+ * ready.
+ */
+Blaze.TemplateInstance.prototype.subscriptionsReady = function () {
```
It seems like this will often trigger deps changed without a value change.  eg, on unsub, it will stay true.

 what's the function that you can use to make that not happen?

stubailo replied:
Fixed this

glasser commented on packages/blaze/view.js:
```diff
@@ -188,6 +188,40 @@ Blaze.View.prototype.autorun = function (f, _inViewScope) {
+
+  self._errorIfShouldntCallSubscribe();
+
+  var subHandle = Meteor.subscribe.apply(Meteor, arguments);
```
would be nice, in theory, to have a way to use a different connection than the default connection.  Not sure how you would work it into this varargs API.  I guess you could just have a different function `subscribeOnConnection`

stubailo replied:
I think we can leave that for a later time until people ask for it?

glasser replied:
Sure, as long as you expect that the answer is going to be something like "add another function with a different name" and not "I sure wish I had designed this function to take an `options` argument but I can't now".

The obvious case to me is — are we going to want to use this behavior in `accounts-ui`?  Because that's going to want to use `Accounts.connection`.

stubailo commented:
Merged manually

KrishnaPG commented:
Looks interesting and would love to use this kind of thing - any examples / tutorials on the how to take advantage of this feature?

stubailo commented:
Here are the official docs: http://docs.meteor.com/#/full/Blaze-TemplateInstance-subscribe

Here is a great article by discover Meteor, written before the feature was added to Meteor: https://www.discovermeteor.com/blog/template-level-subscriptions/
